hpx/components/containers/coarray/coarray.hpp

```diff
@@ -0,0 +1,240 @@
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+///////////////////////////////////////////////////////////////////////////////
+/// \cond NOINTERNAL
+
+#define HPX_REGISTER_COARRAY(T) HPX_REGISTER_PARTITIONED_VECTOR(T)
+
+namespace hpx
+{
+    namespace detail{
+
+        struct auto_subscript
+        {
+            constexpr auto_subscript(){}
+
+            // Defined to pass the coarray_sizes constructor
+            operator std::size_t()
+            {
+                return std::size_t(-1);
+            }
+        };
+
+        template<typename C>
+        struct cast_if_autosubscript
+        {
+            using type = C;
+        };
+
+        template<>
+        struct cast_if_autosubscript<detail::auto_subscript>
+        {
+            using type = int;
+        };
+
+        template<std::size_t N>
+        struct coarray_sizes
+        {
+            using iterator =
+                typename std::array<std::size_t,N>::iterator;
+
+            template<typename ... I>
+            coarray_sizes(I... i)
+            : sizes_({{i...}})
+            {
+                using last_element
+                    = typename hpx::detail::last_element< I... >::type;
+                using condition =
+                    typename std::is_same<last_element,detail::auto_subscript>;
+
+                static_assert(condition::value,
+                    "hpx::coarray() needs the last size to be " \
+                    "equal to hpx::_");
+
+                static_assert(N == sizeof...(I),
+                    "hpx::coarray() needs the number of sizes to be " \
+                    "equal to its dimension");
+
+                static_assert(detail::are_integral<
+                    typename detail::cast_if_autosubscript<I>::type...>::value,
+                        "One or more elements in sizes given to hpx::coarray() "\
+                        "is not integral");
+            }
+
+            iterator begin()
+            {
+                return sizes_.begin();
+            }
+
+            iterator end()
+            {
+                return sizes_.end();
+            }
+
+            template<typename View, typename Iterator, std::size_t ... I>
+            void update_view(
+                View & v,
+                std::size_t num_images,
+                hpx::util::detail::pack_c<std::size_t, I...>,
+                hpx::lcos::spmd_block const & block,
+                Iterator && begin,
+                Iterator && last) const
+            {
+                v = View(block,
+                    std::forward<Iterator>(begin),
+                    std::forward<Iterator>(last),
+                    {(sizes_[I]!= std::size_t(-1) ? sizes_[I] : num_images)...});
+            }
+
+        private:
+            std::array<std::size_t,N> sizes_;
+        };
+
+    }
+
+    // Used for "automatic" coarray subscript and "automatic" coarray size
+    constexpr hpx::detail::auto_subscript _;
```

Comment:
`hpx::coarray` is already use for the class. I momentarily put it in `hpx::container` namespace